Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/CustomControls.fs

```diff
@@ -82,9 +82,7 @@ module BindableHelpers =
                 newHolder.ViewElement.UpdateInherited(prevModelOpt, newHolder.ViewElement, bindableObject)
                 holderOpt <- ValueSome newHolder
                 prevModelOpt <- ValueSome newHolder.ViewElement
-            | _ ->
-                holderOpt <- ValueNone
-                prevModelOpt <- ValueNone
+            | _ -> ()
```

Comment:
@vshapenko This was the main culprit.
Whenever Xamarin.Forms tries to reuse a row, it first set the BindingContext of that row to `null` before setting it the new item.
And since we were resetting the `prevModelOpt` of the row on `null`, when the new item was set into `BindingContext`, Fabulous would think the row was empty and the diffing was incorrect.

Reply:
Sounds brilliant. Never dived so deep.
Some time ago i tried to write an updated version of diff algo. The main idea was to  keep a collection of 

```fsharp
   type Element<'T> =
        | Direct of 'T
        | FromCell of int
 ```
where 'T is TTarget. So, main cycle just fills this list/array/whatever. Second pass maps to proper 'TTarget. and finally we compare elements of two 'TTarget by identical and swap if needed.  Here is the code: 

```fsharp
 type Element<'T> =
        | Direct of 'T
        | FromCell of int
        
    /// Incremental list maintenance: given a collection, and a previous version of that collection, perform
    /// a reduced number of clear/add/remove/insert operations
    let updateCollectionGeneric
           (prevCollOpt: 'T[] voption) 
           (collOpt: 'T[] voption) 
           (targetColl: IList<'TargetT>) 
           (create: 'T -> 'TargetT)
           (attach: 'T voption -> 'T -> 'TargetT -> unit) // adjust attached properties
           (canReuse : 'T -> 'T -> bool) // Used to check if reuse is possible
           (keyOf:'T -> string voption)
           (update: 'T -> 'T -> 'TargetT -> unit) // Incremental element-wise update, only if element reuse is allowed
        =
        let elems =
            match collOpt with
            | ValueSome v->Array.init<Element<'TargetT> option> v.Length (fun _->None)
            | ValueNone -> [||]
         
        let create index child =
            let targetChild = create child
            attach ValueNone child targetChild
            elems.[index]<-Direct targetChild |> Some
            
        let update index prevChild newChild =
            let targetChild = targetColl.[index]
            update prevChild newChild targetChild
            attach (ValueSome prevChild) newChild targetChild
            elems.[index]<-Direct targetChild |> Some
            
        let move prevIndex newIndex =
            elems.[newIndex]<-FromCell prevIndex |> Some
        
        updateCollectionGenericInternal
            prevCollOpt
            collOpt
            keyOf
            canReuse
            (fun () -> targetColl.Clear())
            create
            update
            move
            (fun index -> targetColl.RemoveAt(index))
            
        let newElems =
            [|
                for i in 0..elems.Length-1 do
                    match elems.[i] with
                    | Some (FromCell idx) ->
                        yield targetColl.[idx]
                    
                    |Some (Direct el) ->
                        yield el
                    
                    |None ->
                        yield targetColl.[i]
                    
            |]
        
        for i in 0..newElems.Length-1 do
            if(i>=targetColl.Count) then
                targetColl.Add newElems.[i]
            else 
                if(not (identical newElems.[i] targetColl.[i]))
                 then
                     targetColl.RemoveAt(i)
                     targetColl.Insert(i,newElems.[i])
```

Main idea is to simplify (and minimize) operations with target coll. Also in such implementation currState is not needed, as we can see, because we do not change collection before we have full list of updates